### ISSUE TITLE
NeedBraces: false positives for single line 'if' in 'else', with allowSingleLineStatement enabled

### ISSUE THREAD

**Issue by dhendriks** [checkstyle/checkstyle#1007]
This issue should be considered together with #1006 .

http://checkstyle.sourceforge.net/config_blocks.html#NeedBraces

For the NeedBraces check, with allowSingleLineStatement, I've identified what appear to be a false positive.

Problems are reported for the following: 

``` java
        int x = 0;
        if      (x == 0) x = 1;
        else if (x == 1) x = 2;
        else if (x == 2) x = 3;
        else             x = 4;
```

On the 3rd and 4th lines, it reports that the 'if' needs braces, while it is clearly single line. It seems to me this is a bug, as the 'if' is single line.
##

**Comment by Vladlis** [comment]
I'm working on it

**Comment by Vladlis** [comment]
Well, as I see from code and [input](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/InputBracesSingleLineStatements.java#L29) the example above is considered as complex blocks, whereas this code will pass with no violations:
```java
        int x = 0;
        if      (x == 0) x = 1;
        else {if (x == 1) x = 2;
        else {if (x == 2) x = 3;
        else             x = 4;}}
```
So, we have to decide, whether to allow all complex blocks, which may make such constructions as this one possible:
```java
   if(x == 0) if (x == 1) if (x == 2) if (x == 3) break;
```
or to allow only particular constructions.

I propose a new boolean option, which will allow complex blocks.

**Comment by dhendriks** [comment]
In my opinion, the following should be OK:

```java
    int x = 0;
    if      (x == 0) x = 1;
    else if (x == 1) x = 2;
    else if (x == 2) x = 3;
    else             x = 4;
```

While the following should NOT be ok, as it is missing braces for multiline 'else if' part:

```java
    int x = 0;
    if      (x == 0) x = 1;
    else if (x == 1) 
                     x = 2;
    else if (x == 2) x = 3;
    else             x = 4;

```

I'm not sure what you're proposed new 'allow complex blocks' option means for these two cases. Could you elaborate on that?

**Comment by dhendriks** [comment]
As far as your example is concerned:

```java
    if(x == 0) if (x == 1) if (x == 2) if (x == 3) break;
```

In my opinion, it should be OK, as the 'if' statements are on a single line, and thus doesn't need braces.

**Comment by dhendriks** [comment]
I just realized my previous comments may not have been clear enough. I think we don't need a new 'allow complex blocks' option. The check should just report the proper thing regardless of complex/non-complex ifs. I understand the implementation may need to be different for both case, but for me the definition of multi-line doesn't depend on the 'if' being complex or non-complex. Or am I missing something here?

**Comment by Vladlis** [comment]
>  but for me the definition of multi-line doesn't depend on the 'if' being complex or non-complex. 

Yeah, that's what I'm talking about - if you are sure, that the the code you write won't be confusing for others, then you'll be able to tell the Check not to worry about it and allow such code by setting this newOption to 'true'.
However, there is another question: should we allow only complex 'if' or also allow complex 'for'(#1006), 'while' etc.
Exmple of complex 'for' loop:
```java
     for (int i = 0; i < 10; i++) if (i > 0) for (int j = 0; j < 10; j++) if (j > 0) break;
```
I'm not sure, that the last should be allowed by any option :)

**Comment by romani** [comment]
Related issues, discussions: #300, #895, #715 

@ liutikas , can you share your opinion on code examples that are shown in this issue. https://github.com/checkstyle/checkstyle/issues/1007#issuecomment-120326457

**Comment by romani** [comment]
nested IF or any other structures will not be suppressed - user need to switch OFF Check at all to skip such violations.

ELSE-IF - is conner case, I am almost ok to allow that by existing option.

**Comment by dhendriks** [comment]
```java
for (int i = 0; i < 10; i++) if (i > 0) for (int j = 0; j < 10; j++) if (j > 0) break;
```

As far as this example is concerned, I agree that such for/if/for/if/break nesting might be a bit too much. My suggestion would be to add a new MaximumNestedStatementsOnSingleLine check or so. For instance, with a maximum level of 2, the following (which has 3 levels, for/if/for) is not allowed:

```java
for (int i = 0; i < 10; i++) if (i > 0) for (int j = 0; j < 10; j++) break;
```

while with a maximum level of 3, it will be allowed. This allows users to decide for themselves how many levels they find acceptable. I propose to separate this discussion and this check, from the NeedBraces check, as it seems to be about nesting of statements. If you disallow such nestings on a single line, then they need to be split to multiple lines, and the NeedBraces check will take care of the needed braces in that case. It seems to be different issues/checks.

I would suggest to move the discussion of such 'nested statements on a single line' (I like this name better than the vague 'complex statements' or somewhat ambiguous 'multi level single line statements') to a different issue. This issue is about fixing the false positive, where if the statement is on a single line, and the 'allowSingleLineStatement' option is enabled, it should not report warnings.

**Comment by romani** [comment]
@dhendriks, please move MaximumNestedStatementsOnSingleLine to separate issue .

@dhendriks, please provide more examples of such code format and please with whole logic to let us validate it is really the best form in that cases. The more examples the better. I need to see that such code is reasonable in lots of places. If it is very rare, I would suggest to use suppression.

Even I (personally) do not like such code (as it looks like switch candidate, ...... ):
```
    int x = 0;
    if      (x == 0) x = 1;
    else if (x == 1) x = 2;
    else if (x == 2) x = 3;
    else             x = 4;
``` 
I am ok to allow that if community will vote on it and I will see a demand .


Originally allowSingleLineStatement was created to allow very simple IF structures - #300 .
Original author of that idea confirmed expected violation on that case - https://github.com/checkstyle/checkstyle/issues/300#issuecomment-120565558 .

So we have conflicts of request.  I personally do not like both, but ok with current implementation of option.

@liutikas , it would be good if you share your opinion on that issue.

**Comment by liutikas** [comment]
My initial intent when I requested this feature "allowSingleLineStatement" was to allow for simple one like if statements such as:
```java
if (x == 0) return foo();
```
However, I do not think that code snippet mentioned in this issue should be allowed:
```java
if      (x == 0) x = 1;
else if (x == 1) x = 2;
else if (x == 2) x = 3;
else             x = 4;
```

This is a style that we use for the Chromium project (style that is very close to Google Java style) and I would like to keep this option as it is or at least make it that modifications to "allowSingleLineStatement" would still make it possible to flag else statements that on the same line as that is against our style.

**Comment by dhendriks** [comment]
Here are some examples from our source code of 'if' after 'else' with statement following it on a single line. There may be more of those in our source code, but these should be sufficient for the discussion. I'll comment on each of them.

1)

```java
    public static EDataType makeDataType(String name) {
        EcorePackage ePkg = getFactory().getEcorePackage();
        if (name.equals("EBoolean"))              return ePkg.getEBoolean();
        else if (name.equals("EBooleanObject"))   return ePkg.getEBooleanObject();
        else if (name.equals("EByte"))            return ePkg.getEByte();
        else if (name.equals("EByteObject"))      return ePkg.getEByteObject();
        else if (name.equals("EByteArray"))       return ePkg.getEByteArray();
        else if (name.equals("EChar"))            return ePkg.getEChar();
        else if (name.equals("ECharacterObject")) return ePkg.getECharacterObject();
        else if (name.equals("EDouble"))          return ePkg.getEDouble();
        else if (name.equals("EDoubleObject"))    return ePkg.getEDoubleObject();
        else if (name.equals("EFloat"))           return ePkg.getEFloat();
        else if (name.equals("EFloatObject"))     return ePkg.getEFloatObject();
        else if (name.equals("EInt"))             return ePkg.getEInt();
        else if (name.equals("EIntegerObject"))   return ePkg.getEIntegerObject();
        else if (name.equals("EJavaObject"))      return ePkg.getEJavaObject();
        else if (name.equals("ELong"))            return ePkg.getELong();
        else if (name.equals("ELongObject"))      return ePkg.getELongObject();
        else if (name.equals("EShort"))           return ePkg.getEShort();
        else if (name.equals("EShortObject"))     return ePkg.getEShortObject();
        else if (name.equals("EString"))          return ePkg.getEString();
        else {
            Assert.fail(fmt("Unknown standard data type \"%s\".", name));
            return null; // Not reached.
        }
    }
```

This code is from before Java allowed switch over string typed values. It could now probably be changed to a switch. However, there are probably other such constructs that you can imagine that are not about int/string/enum etc and thus can't be rewritten to a switch. I should also note that we could write this as a bunch of separate 'if' statements, not using 'else' at all, as they all 'return' anyway. However, if this were part of a larger method, with code following it, that would not be possible.

2)

```java
    public String toString(Set<Set<Location>> superLocs) {
        List<String> results = list();
        for (Set<Location> locs: superLocs) {
            List<String> r = list();
            for (Location l: locs) {
                if      (l == loc1) r.add("1");
                else if (l == loc2) r.add("2");
                else if (l == loc3) r.add("3");
                else if (l == loc4) r.add("4");
                else                r.add("?");
            }
            Collections.sort(r, Strings.SORTER);
            String s = "";
            for (String ss: r) {
                if (!s.isEmpty()) s += ",";
                s += ss;
            }
            results.add("{" + s + "}");
        }
        Collections.sort(results, Strings.SORTER);
        String s = "";
        for (String ss: results) {
            if (!s.isEmpty()) s += ",";
            s += ss;
        }
        return "{" + s + "}";
    }
```

As mentioned for (1), there can be other variants that can't be changed to a 'switch' statement. This is one of them. Each location 'l' is checked against some global instances of the 'Location' type, namely 'loc1', 'loc2', 'loc3', and 'loc4'. I can't think of a equally or better readable way to write this, not using 'else if (condition) statement' on a single line.

**Comment by romani** [comment]
>```
        if (name.equals("EBoolean"))              return ePkg.getEBoolean();
        else if (name.equals("EBooleanObject"))   return ePkg.getEBooleanObject();
        else if (name.equals("EByte"))            return ePkg.getEByte();

```

single line is ok only due to similarity of code .  Right now Check is not ready for that analysis.
We might consider to make some experiment in https://github.com/sevntu-checkstyle/sevntu.checkstyle and let it be tested on wide range of sources to prove that extra analysis does not give false-positives ....... and gain maturity.


> ```
                if      (l == loc1) r.add("1");
                else if (l == loc2) r.add("2");
                else if (l == loc3) r.add("3");
                else if (l == loc4) r.add("4");
                else                r.add("?");
```

How to distinguish simple code from complicated code :) automatically ? 
How to catch similarity of code automatically ... ?
That is topic for investigation . 
I personally do not like this as it is hard to debug. Example: sometime you nee to debug a method from JDK , you step in , but IDE is not able to show you values of variables (sometime it happens) , where is no other way to figureout what is going on as to see how interpretor is going from line to line. Problems sometime are so not obvious so it is good to see that interpretor recognized IF as TRUE and about to execute statement (may be I am the only unlucky engineer and experienced strange interpretor behaviour sometime ,may be I am old school).


Summary:
1) Use suppressions , that would be very convenient to use , but not merged yet - https://github.com/checkstyle/checkstyle/issues/1340.
2) Try to implement extension for this Check in experimental project with options to recognize patterns described above, and use it right away instead of basic Check. As you gather more votes on that issue and implementation is ready in experimental project we could continue discussion and merging your experiment to main code.

**Comment by romani** [comment]
closed in favor of summary issue - #3618 

@dhendriks , @Vladlis , @liutikas  please provide your feedback in new issue if you still care.

**Comment by romani** [comment]
request to update to documentation for current behavior - #3619 .
